spring-boot-admin-server-ui/src/main/frontend/services/startup-actuator.js

```diff
@@ -0,0 +1,143 @@
+ * limitations under the License.
+ */
+
+const regex = new RegExp('([^=\\s]*)=\\[([^\\]]*)\\]', 'gi')
+
+export class StartupActuatorEventTree {
+  constructor(events) {
+    this.events = events;
+  }
+
+  getEvents() {
+    return this.events || [];
+  }
+
+  getRoots() {
+    return this.getByDepth(0);
+  }
+
+  getByDepth(depth) {
+    return this.getEvents().filter((event) => event.startupStep.depth === depth)
+  }
+
+  getById(id) {
+    return this.getEvents().find((event) => event.startupStep.id === id)
+  }
+
+  getByParentId(parentId) {
+    return this.getEvents().filter((event) => event.startupStep.parentId === parentId)
+  }
+
+  getStartTime() {
+    return this.getEvents().map(e => Date.parse(e.startTime)).reduce((a, b) => Math.min(a, b), Number.MAX_VALUE)
+  }
+
+  getEndTime() {
+    return this.getEvents().map(e => Date.parse(e.endTime)).reduce((a, b) => Math.max(a, b), Number.MIN_VALUE)
+  }
+
+  getPath(id) {
+    let event = this.getById(id);
+    if (!event) {
+      return [];
+    }
+
+    const path = [id]
+    let parent = event.startupStep.parent;
+    while (parent !== null && parent !== undefined) {
+      path.push(parent.startupStep.id);
+      parent = parent.startupStep.parent;
+    }
+
+    return path;
+  }
+
+  getPeriod(event) {
+    const eventStartTime = Date.parse(event.startTime);
+    const eventEndTime = Date.parse(event.endTime)
+    const treeStartTime = this.getStartTime();
+    const treeEndTime = this.getEndTime();
+
+    const treeTimeSpan = treeEndTime - treeStartTime
+    const relativeEventStartTime = eventStartTime - treeStartTime;
+    const relativeEventEndTime = eventEndTime - treeStartTime;
+    const relativeStart = relativeEventStartTime > 0 ? relativeEventStartTime / treeTimeSpan : 0;
+    const relativeEnd = relativeEventEndTime > 0 ? relativeEventEndTime / treeTimeSpan : 0;
+
+    return {
+      start: relativeStart,
+      end: relativeEnd,
+    };
+  }
+}
+
+export const StartupActuatorService = {
+  parseAsTree(data) {
+    let events = data.timeline.events || [];
+    let eventsForTree = events
+      .sort((a, b) => a.startupStep.id - b.startupStep.id)
+      .map((event) => {
+        event.startupStep.parent = this.getById(events, event.startupStep.parentId);
+        event.startupStep.tags = event.startupStep.tags.map(this.parseTag)
+        event.startupStep.depth = 0;
+        event.startupStep.children = this.getByParentId(events, event.startupStep.id);
+        event.duration = this.convertToMicroseconds(event.duration);
+        return event;
+      })
+      .map((event) => {
+        let parent = event.startupStep.parent;
+        while (parent !== null && parent !== undefined) {
+          parent = parent.startupStep.parent;
+          event.startupStep.depth++;
+        }
+
+        return event;
+      });
+
+    return new StartupActuatorEventTree(eventsForTree);
+  },
+  convertToMicroseconds: function (duration) {
```

Comment:
I wonder if this function catches the different values that the `duration` parameter can have.
The startup endpoint responds a serialized Java `Duration`: https://github.com/spring-projects/spring-boot/blob/769b5f0f03bd34e1d18d7b0ec54f0fcdacf0fd03/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/context/metrics/buffering/StartupTimeline.java#L76 

Should the function here handle `Duration`s that are of greater value. 
The format could be (see https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html#toString--):
`PT20.345S`  (equals 20.345 seconds)

Although the startup of an application is (or should) likely to be relatively fast. ;-)

Is it worth it to use some library for parsing the duration String? For example https://www.npmjs.com/package/iso8601-duration?

Regarding naming of the function: Is the returned valued of unit microseconds or milliseconds?

Reply:
Now supports up to any amount of weeks. I assume that a startup step does never take so long. The displayed value is in milliseconds.

Reply:
Sorry for being this late with my review, @SteKoe! 🙏🏻 
This looks already good to me.
The iso8601-duration library's example shows a `toSeconds` function:
```javascript
toSeconds( parse('PT1H30M10.5S')
```
Do you think that function can cover the converting part in our `toMilliseconds` function replacing all multiplications but the last?